accepted/0019-app-notifications.md

```diff
@@ -0,0 +1,70 @@
+
+Note: This Alert system would not expose a way to pop up warning/error dialogs. That is not its purpose.
+
+## Internals
+
+The Alerts framework would likely have to be built as part of libobs, providing a pathway from plugins to the UI.
+
+When generating an Alert, the following (bare minimum) parameters should be available:
+
+```js
+{
+  "timestamp": 1585817608, // An epoch timestamp of when the issue occurred. Generated automatically
+  "priority": "CRITICAL", // Enum of CRITICAL, WARNING, INFO // Required? Could default to warning?
+  "summary": "", // Required. HTML-anchor capable. A short summary of the issue, visible from the main window
+  "description": "", // Optional. A lengthy, more informative HTML string providing more context and inline solutions. Might support newlines and dot points
+  "persist": true, // Optional, default to false. If the user dismisses the alert, whether it should persist in the full list. Useful for alerts that are tied to a fixable issue.
+}
+```
+
+Creating this Alert should return a unique identifier, so that the plugin can keep track and control the status of an Alert. Alongside this, we should also provide an ability to tie Alerts to a specific source (scene? group?), property, and its valid state. That way the plugin relinquishes control of the Alert and it's dismissed as part of the property becoming valid, or the source being deleted/unloaded (when switching Scene collections or closing OBS).
+
+# How We Teach This
+
+On first launch, it'd be nice to show an initial bit of information via this system to introduce the user to OBS, and to explain the purpose of the Alert UI.
+
+For developers, making sure this is well documented is important. Knowing that they can reach out to the user within the UI when things go wrong is important.
+
+# Drawbacks
+
+Complexity. This will require a lot of code, logic, and a background runner keeping track of active/dismissed alerts. The solution to this may be the option to turn it off somehow for bigger productions that only need issues confirmed before going live.
+
+Visual noise. This one is less of a concern, but at the same time we don't want too much going on in the UI. A big toolbar along the top could be distracting, but at the same time going subtle means that the user could potentially not realise that something has gone wrong.
+
+# Additional Information
+
+There are many things to consider in this design. It's highly likely that the backend & UI would be developed & PR'd independently rather than together. This'd ensure a solid backend design, and an easier time reviewing it.
+
+The backend design & scope should be fleshed out significantly more before a real impelementation is started.
+
+# Unresolved questions
+
+* Should Alerts have a timeout? A list of methods to dismiss (scene collection switch, profile switch)?
```

Comment:
We can probably "solve" that second part to some degree by following the split that e.g. the hotkey system follows: `obs_add_source_notification`/`obs_add_frontend_notification`/`obs_add_output_notification`

Neither scene collection nor profile are libobs concepts, they're UI concepts, so exposing those in libobs is not a good idea imo. Instead what libobs can do is dismiss notifications from sources/outputs/services/… that get removed

So for example an image source notification "could not open file xyz.jpg" would go away automatically once that source is removed

I can't think of examples of where I'd want notifications to have a timeout, and the more I think about it the more I'd want all notifications to be logged (to prevent the whole "phone made a notification sound while nothing whatsoever was going on, but when checking the phone it didn't display anything, and there was no record of what had happened"), but this is likely going to cause fatigue issues very fast if there's no "sane" bulk dismiss option